src/BuildIt.Forms/BuildIt.Forms.Controls/StatefulControl.xaml.cs

```diff
@@ -11,22 +13,99 @@ namespace BuildIt.Forms.Controls
     [XamlCompilation(XamlCompilationOptions.Compile)]
     public partial class StatefulControl
     {
+        public static readonly BindableProperty EmptyStateTemplateProperty = BindableProperty.Create(nameof(EmptyStateTemplate), typeof(DataTemplate), typeof(StatefulControl), propertyChanged: HandleEmptyStateTemplateChanged);
+
+        // TODO MK Disabling pull to refresh by default, as it's still a WIP
+        public static readonly BindableProperty IsPullToRefreshEnabledProperty = BindableProperty.Create(nameof(IsPullToRefreshEnabled), typeof(bool), typeof(StatefulControl), defaultValue: false);
+        public static readonly BindableProperty LoadingErrorStateTemplateProperty = BindableProperty.Create(nameof(LoadingErrorStateTemplate), typeof(DataTemplate), typeof(StatefulControl), propertyChanged: HandleLoadingErrorStateTemplateChanged);
+        public static readonly BindableProperty LoadingStateTemplateProperty = BindableProperty.Create(nameof(LoadingStateTemplate), typeof(DataTemplate), typeof(StatefulControl), propertyChanged: HandleLoadingStateTemplateChanged);
+        public static readonly BindableProperty PullToRefreshBackgroundColorProperty = BindableProperty.Create(nameof(PullToRefreshBackgroundColor), typeof(Color), typeof(StatefulControl), propertyChanged: HandlePullToRefreshBackgroundColorPropertyChanged);
+        public static readonly BindableProperty PullToRefreshCommandProperty = BindableProperty.Create(nameof(PullToRefreshCommand), typeof(ICommand), typeof(StatefulControl));
+        public static readonly BindableProperty PullToRefreshStateErrorTemplateProperty = BindableProperty.Create(nameof(PullToRefreshStateErrorTemplate), typeof(DataTemplate), typeof(StatefulControl), propertyChanged: HandlePullToRefreshErrorStateTemplateChanged);
+        public static readonly BindableProperty PullToRefreshStateTemplateProperty = BindableProperty.Create(nameof(PullToRefreshStateTemplate), typeof(DataTemplate), typeof(StatefulControl), propertyChanged: HandlePullToRefreshStateTemplateChanged);
         public static readonly BindableProperty StateProperty = BindableProperty.Create(nameof(State), typeof(StatefulControlStates), typeof(StatefulControl), propertyChanged: HandleStatePropertyChanged);
 
-        private const string EmptyStateContainerName = "EmptyStateContainer";
-        private const string ErrorStateContainerName = "ErrorStateContainer";
         private const uint FadeInAnimationTimeInMilliseconds = 250;
         private const uint FadeOutAnimationTimeInMilliseconds = FadeInAnimationTimeInMilliseconds / 2;
         private const double FullyOpaque = 1;
         private const double FullyTransparent = 0;
-        private const string LoadingStateContainerName = "LoadingStateContainer";
+        private const int MaxStatesHistoryEntries = 20;
+
+        // MK Making sure that these callbacks won't get GC'd --> https://stackoverflow.com/a/51466538/510627
+        private readonly Func<bool> canPullToRefresh;
```

Comment:
@nickrandolph Please have a look at this solution to use WeakReference in order to prevent the control from causing memory leaks. I'm not sure if this is the best option.